opte/src/engine/packet.rs

```diff
@@ -1707,11 +1840,60 @@ impl PacketSeg {
         self.len -= amount;
     }
 
-    fn slice(&mut self, off: usize, len: usize) -> &[u8] {
-        unsafe {
-            let start = (*self.mp).b_rptr.add(off);
-            slice::from_raw_parts(start, len)
-        }
+    /// Get a slice of the entire segment.
+    fn slice(&self) -> &[u8] {
+        // Safety: We are using the segment's own data to take a slice
+        // of the entire segment.
+        unsafe { self.slice_unchecked(0, None) }
+    }
+
+    /// Get a mutable slice of the entire segment.
+    fn slice_mut(&mut self) -> &mut [u8] {
+        // Safety: We are using the segment's own data to take a slice
+        // of the entire segment.
+        unsafe { self.slice_mut_unchecked(0, None) }
+    }
+
+    /// Get a slice of the segment.
+    ///
+    /// The slice starts at `offset` and consists of `len` bytes. If
+    /// the length is `None`, then the slice extends to the end of the
+    /// segment. This includes only the part of the dblk which has
+    /// been written, i.e. the bytes from `mblk.b_rptr` to
+    /// `mblk.b_wptr`.
+    ///
+    /// # Safety
+    ///
+    /// It is up to the caller to ensure that `offset` and `offset +
+    /// len` reside within the segment boundaries.
+    unsafe fn slice_unchecked(
+        &self,
+        offset: usize,
+        len: Option<usize>,
+    ) -> &[u8] {
+        let start = (*self.mp).b_rptr.add(offset);
+        slice::from_raw_parts(start, len.unwrap_or(self.len))
```

Comment:
The length argument to `from_raw_parts` seems wrong, unless I'm missing something. I would expect that to be something like `len.unwrap_or(self.len - offset)`. If I provide arguments of `1` and `None`, then the slice will extend beyond the limit of the underlying `mblk`, right?

Also, that subtraction could be something like `self.len.saturating_sub(offset)` to ensure that we never panic on underflow. We'd rather get an empty slice I think.

Reply:
You are correct, this was wrong. I dabbled with your idea, but we also had the issue that offset could be past the end of the segment. Then I dabbled with the idea of silently pinning `offset` to `b_wptr` when its too large, and pinning `offset + len` to `b_wptr` when it goes past, effectively making the function always safe...but then I realized this is a bad idea because it's no longer correct, and it's pretty unsporting for it to silently reduce the size of the expected slice (and this would also result in very silent/hard-to-debug issues in production). So yea, I opted for panic. If we can to eliminate all possibility of panic either the parsing/packet code needs to change to avoid this slicing all together or we make a version that returns Option/Result and deal with that.

Reply:
Yeah, those are both good points, that we have to deal with limiting original offset and silently returning something "else" sounds bad. That does lead me to wonder though, why not make this return an optional type now? It looks like the only callsites that do _not_ currently have to handle errors are the `Packet::slice` and `Packet::slice_mut` functions, which are implemented in terms of `slice_unchecked`. For example, `Packet::body_segs` already returns an `Option`, so we could just propagate that if the slice fails. Same thing for `Packet::emit_headers`.

I'm not necessarily advocating we do that -- I know this PR has already dragged on much longer than you want to. But I know we've talked about not introducing further panics if possible. I guess one argument is that this really is a bug: if we can't take the slice we ask for, we've parsed the packet incorrectly. Anyway, just want to make sure we're on the same page about that tradeoff, adding a new possible panic site in exchange for not dealing with an optional value.

Reply:
The problem is the other call sites of `slice_unchecked()` like `emit_headers()`, they are currently not prepared to deal with Result/Option from that call. And really they NEVER should be slicing outside the bounds of the segment, because they are slicing based on the `Parsed` data of the packet. So it just didn't seem like a can of worms to break open in this PR. Yes, we could add a new WriteError value and propagate up from new "unexpected" error or something (or maybe it could reuse OutOfRange)...anyways I just didn't want to have to think about all this in an already large PR.

And yea I wrote `slice`/`slice_mut` that way because they are literally just asking the segment for the entire buffer, using the pointers the segment itself deems as the bounds.